supervisor/qemu/src/main.rs

```diff
@@ -989,6 +999,47 @@ impl QemuSupervisor {
             qemu_proc,
         });
     }
+
+    async fn qemu_monitor_shutdown(qmp_socket_path: &str) -> Result<(), anyhow::Error> {
+        let mut stream = TcpStream::connect(qmp_socket_path).await?;
```

Comment:
Does the `TcpStream` here also work with UNIX sockets? The invocation of this function in 1221 suggests that this is called with a UNIX socket path instead? Is there a more appropriate stream type to use here?

Reply:
I was actually just going to take this PR down for similar reasons that after thinking about it I don't think this is a very extensible setup. I propose we switch to using the qapi-qmp crate for QMP communication instead of hard coded socket paths. This will address the socket type issues, and also allow for more extensibility (probably we need to use it for snapshots). I'll close this PR and update the PR to implement this approach if you think this is a good idea?

Reply:
I don't really have any experience working with the `qapi-qmp` crate, but I'm not opposed to it. We'll need to make sure that we don't block our async executor, so either the crate will have to be async or we wrap calls into a `tokio::task::block_in_place`.

Generally using a dedicated API crate sounds like a good idea to me. Whether it's worth the effort probably depends on how deep our integration into QEMU will be: for instance, if we take live snapshots, that'll be another monitor command, whereas if we only do offline snapshots, that's just a `qemu-img` command invocation.

Reply:
If we migrate to this crate, we should still adjust this logic to avoid locking the job mutex for a long period of time: https://github.com/treadmill-tb/treadmill/pull/25#discussion_r1701939086